processing/src/main/java/org/apache/druid/data/input/ProjectionSpec.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import org.apache.druid.data.input.impl.AggregateProjectionSpec;
+import org.apache.druid.query.OrderBy;
+import org.apache.druid.segment.VirtualColumns;
+
+import java.util.List;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+    @JsonSubTypes.Type(name = "aggregate", value = AggregateProjectionSpec.class)
+})
+public interface ProjectionSpec
```

Comment:
The other interface I had in mind is projections which are not using a rollup facts holder, which came from some offline discussions I have had with @gianm, but I have not built this yet. The thinking was that since we've lifted the restriction that segments have to be time ordered, there seems likely to be some use cases for projections which filter, transform, or order the data differently than what the 'base' table does but without grouping the data. I suspect we would always want these to be filtered, or else represent them in persisted segments quite a bit differently (maybe just an index for ordering with no filtering).

The ordering stuff is also not fully realized yet so maybe the utility of these kinds of projections isn't quite apparent at this time. We were thinking that query engines should be able to take advantage of the cases where `CursorHolder.getOrdering()` is the same as the order specified by the `CursorBuildSpec.getPreferredOrdering()`, but right now its use is primarily limited to making cursors for ascending or descending time ordered data for the timeseries, timeboundary, and scan with time ordering queries (which are the only things which currently set preferredOrdering on the build spec today).

I'll back out the interface for now until we add other implementations.